From ordex@autistici.org Fri Nov 16 16:49:18 2012
X-Apparently-To: lindner_marek@yahoo.de via 188.125.84.189; Fri, 16 Nov 2012 08:50:26 +0000
Received-SPF: pass (domain of autistici.org designates 178.255.144.35 as permitted sender)
X-YMailISG: qdh1iTQWLDs62bkFFSythfvhHaZxhfpJ5PSKnjpe7hExOA6w
 0Hc3JpktYyerT5XQmK0XpfEp9b0El16poZFQ_PqNppiGeXVbVwgYE8uvjLm1
 kNwrYEvCc0nHo78mCSQkk_ilI7QHsL0vneWF0aISE9yqNNXjqyQkimfmRRfQ
 U7DiQvok4JxuXMCS9G86op0LQLt0H9_iTX8cZbvEXjERIlpIKSVn7PScRlSw
 1DwoTaTewPT5dANPP9vv5ABNNcAzPsn1krQ1_.bPyorrDCQBWScspYXaxyD5
 WogJDPx5.tpKbyvcx.ke7zNRpChO4Uu5HThhh0y8wZYSWWuo5qqXA82ZHzqT
 9ARlQ7Ipvkb18vDmUaLRuCHvsGQh5rSS_ctctbZpvvReA7kOV1y_V_itOE9X
 TRJxcQZqqavJUZ0xP9ZAxWc_KiB8BFKwBc_acUbjDr7cRtYLkP89nPFESPH_
 zTku3JwOSGw0.s3vGyEqMxTxAuwLfbf.OHLsRH5eaa0FEg2QMvMDJ4rpFTAm
 8AW08y0Yas.Do1MFuaqX__dn0dOyzvh_jABfSg2mcBqO5XCgo4JG4HNdDUWN
 lc8Z82D1zIRm5XZczZa97xi81_bNB8Gut7deqJLasqjpJ4q4S7NpMUaDDC..
 5beC2z.krETtVHhB1JIQl8gOKLgvJBqEcxvxedbqGpV7lbLXSDgqZG.KobYC
 j4jsdXWEJV6hBl.mJggEWlcMSAsVI.ZVPqrtOIcdGhBis3KDAwp2ULlj3Iu4
 hEwgED_.r_adAqcC1GSqAEIMNF4tvvjsN5GvEcwX2hWnEOYFZ6Kp93SilV1l
 D3rXsN_d8M5jVAvbFw7qOYFZ5rCr0LXcNFuclqDJznNsY8xMkhvP0PP53lwU
 bpZEVuO2UzqA.r_4SeBTHy7tcDvYXlQlW4Q16xYT6sZrE8dXVMrgSBlOKq5G
 1o6N6Xkg2I29H0w.vu2KU6b1WAJlMy2H4dcEv3Zooh5VtQmtYwtmbZDHrIQG
 8lJ381Z9jh7pIFCMsp2epxXWbAqkHd6_r6a3lxpHtnHLWI55RWFI5qNxvxQd
 bTIgorpXL.WAmwiLv6MNomPSuHdRy_rHIX9R8DljaWlLKdI_uoa_lbrdjblJ
 iokyPUKRSx_VKHi.eIfz9JaFOTdJQACoxmSoL391XOewRlctIBRDTvB6yOss
 7Nt9dmv0ePFvTJN9euRL
X-Originating-IP: [178.255.144.35]
Authentication-Results: mta1067.mail.ird.yahoo.com  from=autistici.org; domainkeys=neutral (no sig);  from=autistici.org; dkim=pass (ok)
Received: from 127.0.0.1  (EHLO contumacia.investici.org) (178.255.144.35)
  by mta1067.mail.ird.yahoo.com with SMTP; Fri, 16 Nov 2012 08:50:26 +0000
Received: from [178.255.144.35] (contumacia [178.255.144.35]) (Authenticated sender: ordex@autistici.org) by localhost (Postfix) with ESMTPSA id D78F1E878B;
	Fri, 16 Nov 2012 08:50:23 +0000 (UTC)
X-DKIM: Sendmail DKIM Filter v2.8.2 contumacia.investici.org D78F1E878B
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed; d=autistici.org;
	s=stigmate; t=1353055825;
	bh=mBzsqJo806YwPoG1l/528KrliiW3q+mkiMzUs+PZfkI=;
	h=From:To:Cc:Subject:Date:Message-Id:In-Reply-To:References;
	b=RO+bvxnncVKR0YL3Kl8GMbrFsqGikhqR3MzXRrdYewxnNXaSfY55bVw8C9Av108Id
	 0FIk7xS/FYydezTX95Ln+kRCU96uKVktfvi9+3M3k5Ht/pAaAnRr5DkQ8AbE8+NFyT
	 VmQUdEYEqU8x7/DlrMOsrjxyqmA5a5By+75mm5jk=
From: Antonio Quartulli <ordex@autistici.org>
To: davem@davemloft.net
Cc: netdev@vger.kernel.org,
 Simon Wunderlich <simon.wunderlich@s2003.tu-chemnitz.de>,
 Marek Lindner <lindner_marek@yahoo.de>,
 Sven Eckelmann <sven@narfation.org>,
 Antonio Quartulli <ordex@autistici.org>,
 Simon Wunderlich <siwu@hrz.tu-chemnitz.de>
Subject: [PATCH] batman-adv: process broadcast packets in BLA earlier
Date: Fri, 16 Nov 2012 09:49:18 +0100
Message-Id: <1353055758-2901-5-git-send-email-ordex@autistici.org>
X-Mailer: git-send-email 1.8.0
In-Reply-To: <1353055758-2901-1-git-send-email-ordex@autistici.org>
References: <1353055758-2901-1-git-send-email-ordex@autistici.org>
Status: R
X-Status: N
X-KMail-EncryptionState:  
X-KMail-SignatureState:  
X-KMail-MDN-Sent:  

The logic in the BLA mechanism may decide to drop broadcast packets
because the node may still be in the setup phase. For this reason,
further broadcast processing like the early client detection mechanism
must be done only after the BLA check.

This patches moves the invocation to BLA before any other broadcast
processing.

This was introduced 30cfd02b60e1cb16f5effb0a01f826c5bb7e4c59
("batman-adv: detect not yet announced clients")

Reported-by: Glen Page <glen.page@thet.net>
Signed-off-by: Simon Wunderlich <siwu@hrz.tu-chemnitz.de>
Signed-off-by: Antonio Quartulli <ordex@autistici.org>
Signed-off-by: Marek Lindner <lindner_marek@yahoo.de>
---
 net/batman-adv/soft-interface.c | 12 ++++++------
 1 file changed, 6 insertions(+), 6 deletions(-)

diff --git a/soft-interface.c b/soft-interface.c
index b9a28d2..ce0684a 100644
--- a/soft-interface.c
+++ b/soft-interface.c
@@ -325,6 +325,12 @@ void batadv_interface_rx(struct net_device *soft_iface,
 
 	soft_iface->last_rx = jiffies;
 
+	/* Let the bridge loop avoidance check the packet. If will
+	 * not handle it, we can safely push it up.
+	 */
+	if (batadv_bla_rx(bat_priv, skb, vid, is_bcast))
+		goto out;
+
 	if (orig_node)
 		batadv_tt_add_temporary_global_entry(bat_priv, orig_node,
 						     ethhdr->h_source);
@@ -332,12 +338,6 @@ void batadv_interface_rx(struct net_device *soft_iface,
 	if (batadv_is_ap_isolated(bat_priv, ethhdr->h_source, ethhdr->h_dest))
 		goto dropped;
 
-	/* Let the bridge loop avoidance check the packet. If will
-	 * not handle it, we can safely push it up.
-	 */
-	if (batadv_bla_rx(bat_priv, skb, vid, is_bcast))
-		goto out;
-
 	netif_rx(skb);
 	goto out;
 
-- 
1.8.0

